sopel/irc/modes.py

```diff
@@ -0,0 +1,296 @@
+"""Mode management for IRC channels.
+
+.. seealso::
+
+    https://modern.ircdocs.horse/#mode-message
+
+.. versionadded:: 8.0
+"""
+from __future__ import generator_stop
+
+import enum
+import logging
+from typing import Dict, Generator, List, NamedTuple, Optional, Set, Tuple
+
+
+LOGGER = logging.getLogger(__name__)
+
+
+class ParamRequired(enum.Enum):
+    """Enum of param requirement for mode types."""
+    ALWAYS = 'always'
+    """The mode type always requires a parameter."""
+    ADDED = 'added'
+    """The mode type requires a parameter only when the mode is added."""
+    REMOVED = 'removed'
+    """The mode type requires a parameter only when the mode is removed."""
+    NEVER = 'never'
+    """The mode type never requires a parameter."""
+
+
+DEFAULT_MODETYPE_PARAM_CONFIG = {
+    'A': ParamRequired.ALWAYS,
+    'B': ParamRequired.ALWAYS,
+    'C': ParamRequired.ADDED,
+    'D': ParamRequired.NEVER,
+}
+"""Default parameter requirements for mode types."""
+
+
+class ModeException(Exception):
+    """Base exception class for mode management."""
+    pass
+
+
+class ModeTypeUnknown(ModeException):
+    """Exception when a mode's type is unknown or cannot be determined."""
+    def __init__(self, mode) -> None:
+        super().__init__('Unknown type for mode %s' % mode)
+
+
+class ModeTypeImproperlyConfigured(ModeException):
+    """Exception when the mode's type management is not configured properly."""
+    def __init__(self, mode: str, letter: str) -> None:
+        message = 'Type {mode} for mode {letter} is not properly configured.'
+        super().__init__(message.format(mode=mode, letter=letter))
+
+
+def _parse_modestring(modestring: str
+                      ) -> Generator[Tuple[str, bool], None, None]:
```

Comment:
I honestly hate this style for no particular reason. I can't tell you why. 😅

I also still think that wrapping at 79 chars is too narrow and we should be using 100 or 120, but eh. Suggestion in case you want to take pity on my sensibilities. It's a whopping 83 cols wide. 😝

```suggestion
def _parse_modestring(modestring: str) -> Generator[Tuple[str, bool], None, None]:
```

Reply:
Another common practice is to do that:

```suggestion
def _parse_modestring(modestring: str
) -> Generator[Tuple[str, bool], None, None]:
```